src/ScaledPolynomialTransformFitter.cc

```diff
@@ -0,0 +1,465 @@
+        catalog,
+        keys,
+        maxOrder,
+        intrinsicScatter,
+        computeScaling(catalog, keys.input),
+        computeScaling(catalog, keys.output)
+    );
+}
+
+ScaledPolynomialTransformFitter ScaledPolynomialTransformFitter::fromGrid(
+    int maxOrder,
+    afw::geom::Box2D const & bbox,
+    int nGridX, int nGridY,
+    ScaledPolynomialTransform const & toInvert
+) {
+    Keys const & keys = Keys::forGrid();
+    afw::table::BaseCatalog catalog(keys.schema);
+    catalog.reserve(nGridX*nGridY);
+    afw::geom::Extent2D dx(bbox.getWidth()/nGridX, 0.0);
+    afw::geom::Extent2D dy(0.0, bbox.getHeight()/nGridY);
+    for (int iy = 0; iy < nGridY; ++iy) {
+        for (int ix = 0; ix < nGridX; ++ix) {
+            afw::geom::Point2D point = bbox.getMin() + dx*ix + dy*iy;
+            auto record = catalog.addNew();
+            record->set(keys.output, point);
+            record->set(keys.input, toInvert(point));
+        }
+    }
+    return ScaledPolynomialTransformFitter(
+        catalog,
+        keys,
+        maxOrder,
+        0.0,
+        computeScaling(catalog, keys.input),
+        computeScaling(catalog, keys.output)
+    );
+}
+
+ScaledPolynomialTransformFitter::ScaledPolynomialTransformFitter(
+    afw::table::BaseCatalog const & data,
+    Keys const & keys,
+    int maxOrder,
+    double intrinsicScatter,
+    afw::geom::AffineTransform const & inputScaling,
+    afw::geom::AffineTransform const & outputScaling
+) :
+    _keys(keys),
+    _intrinsicScatter(intrinsicScatter),
+    _data(data),
+    _outputScaling(outputScaling),
+    _transform(
+        PolynomialTransform(maxOrder),
+        inputScaling,
+        outputScaling.invert()
+    ),
+    _vandermonde(data.size(), detail::computePackedSize(maxOrder))
+{
+    // Create a matrix that evaluates the max-order polynomials of all the (scaled) input positions;
+    // we'll extract subsets of this later when fitting to a subset of the matches and a lower order.
+    for (std::size_t i = 0; i < data.size(); ++i) {
+        afw::geom::Point2D input = getInputScaling()(_data[i].get(_keys.input));
+        // x[k] == pow(x, k), y[k] == pow(y, k)
+        detail::computePowers(_transform._poly._u, input.getX());
+        detail::computePowers(_transform._poly._v, input.getY());
+        // We pack coefficients in the following order:
+        // (0,0), (0,1), (1,0), (0,2), (1,1), (2,0)
+        // Note that this lets us choose the just first N(N+1)/2 columns to
+        // evaluate an Nth order polynomial, even if N < maxOrder.
+        for (int n = 0, j = 0; n <= maxOrder; ++n) {
+            for (int p = 0, q = n; p <= n; ++p, --q, ++j) {
+                _vandermonde(i, j) = _transform._poly._u[p] * _transform._poly._v[q];
+            }
+        }
+    }
+}
+
+void ScaledPolynomialTransformFitter::fit(int order) {
+    if (order < 0) {
+        order = _transform.getPoly().getOrder();
+    }
+    int const packedSize = detail::computePackedSize(order);
```

Comment:
If order is greater than the max order computed I think it will read off the end of an array. I suggest a test here with a useful exception that includes the max allowed and the value provided. That means you always must call `_transform.getPoly().getOrder()` which suggests setting a new variable maxOrder to that, then modifying the test above to use maxOrder. That would clarify the `if (order < 0)` test a bit, as well.